### PR TITLE
Add Forvo service (including API key)

### PR DESCRIPTION
**Pull request by Dador**
Hi!

Many people wanted to see Forvo in AwesomeTTS. Me too. So here it is!

I used API key from official Forvo Android application so limit shouldn't be a problem.

I wanted to add setting for API key, but then decided that user can accidentally delete default value. Also there is no much sense in specifying API key if default key works for everyone.

P.S. Sorry if there is style errors, Python isn't my native language 😄

Also, for linking purposes: issue #7.

### CODE REVIEW

**Comment by felixraumann** [comment]
it is probably better to have a version that lets the user enter the key and hostname to use (like ispeech)

**Comment by Dador** [comment]
> it is probably better to have a version that lets the user enter the key and hostname to use (like ispeech)
Well, current API key is almost unlimited (because it's from official Android application and they can't just block it). Field for api key and domain can be added but is there even a use case for this?

**Comment by BorisMonth** [comment]
Since I just started taking a Farsi (Persian) class and there is not TTS available for this language except for eSpeak, which is very bad, I merged your (@Dador) code to my local copy of Anki manually. Everything works and I'm more than happy. It would be nice if it got merged soon, so everyone can use it ;) Thank you very much for your work Dador.

And since I don't know where to put this:
Forvo is a dictionary, better than TTS, and I love it. But as a fall back it would be nice to have a TTS service that's not as bad as eSpeak. Not even Google does provide TTS for Farsi. But I discovered this website right here, which produces very nice sounding Farsi:
http://farsireader.com/webdemoen/
I'll not do it myself, but I would like to leave this message here in case someone is interested in maybe somehow integrating it into AwesomeTTS.

**Comment by Vaanen** [comment]
Hello, so this is very interesting, however i may be a noob and i don't know how to add the support. What do i need to do ? I searched everywhere for any awesomeTTS related file on my computer and didn't find any :(

**Comment by BorisMonth** [comment]
Hello Vaanen,
I'm sorry for not responding. I didn't check my email in a while.
I'm not the author of the project, but your right, maybe I should provide a
description on how to do it, but I don't know if git hub is the right way
to do that. but here you go:

Frovo speak dictionary integration:

https://forvo.com/languages/fa/

Is very nice and apparently someone already implemented it for AwesomeTTS:

https://github.com/AwesomeTTS/awesometts-anki-addon/pull/10

The repository owner didn’t merge it yet, but since it’s python you can
merge it into your own local copy manually …

First you need to find your Anki folder. Here the documentations tells you
where to find it:

https://apps.ankiweb.net/docs/manual.html#files

For me on my Debian system it is on /home/boris/Documents/Anki/

Now copy this folder, just as a backup, just in case something goes wrong…

Go here:

https://github.com/AwesomeTTS/awesometts-anki-addon/pull/10/files

Basically you have to replace the three files in your Anki folder with the
new ones you find here. Here is one way you could do it:


For every of the three files do the following:

   1.

   right click on the “view button” -> Save link as
   2.

   navigate to the AwesomeTTS plugin folder in your anki folder, which for
   me was: /home/boris/Documents/Anki/addons/awesometts/
   3.

   Replace the original file.


This way you replace all three files and you’re done.

Good luck!

**Comment by pcuinty** [comment]
Followed instructions by @BorisMonth and got the following Error:

```
Traceback (most recent call last):
  File "aqt/addons.py", line 41, in loadAddons
  File "/home/user/.local/share/Anki2/addons/AwesomeTTS.py", line 42, in <module>
    import awesometts  # noqa, pylint:disable=wrong-import-position
  File "/home/user/.local/share/Anki2/addons/awesometts/__init__.py", line 7
    <!DOCTYPE html>
    ^
SyntaxError: invalid syntax

```

Anki 2.0.47. on Ubuntu 16.04

**Comment by BorisMonth** [comment]
You are right. Step 1 was wrong. I edited my comment above so now it should work:
" 1. click on the “view button” -> right click on "Raw" button -> Save link as "

**Comment by pcuinty** [comment]
It works great, thank you for your work. 

It would be nice if one could use some of the filters Forvo offers e.g. male/female or filter by country/region.

**Comment by guillaumebrunerie** [comment]
> Field for api key and domain can be added but is there even a use case for this?

Ethics? Using Forvo’s API is a paid service, and rather inexpensive ($1 per month), so it seems very unethical to me to bypass that.

**Comment by shilinglong12138** [comment]
thanks a million. I'm learning Japanese. This helps me a lot.  Leave comments to express my appreciation.😍

**Comment by shilinglong12138** [comment]
it works for 2.0.

**Comment by khonkhortisan** [comment]
@Imrith Yes: from **urllib.parse** import quote

**Comment by Imrith** [comment]
Thank you!

**Comment by Archibalde** [comment]
Hello,

How does the `forvo.py` code deals with the fact that there can be several audio for the same word in a given language? Is there a way to choose within AwesomeTTS interface? The option may have been developed for other services, otherwise I would happily try to add it, even if I am not very familiar with classes in Python and things such as `self`.

If somebody could explain what the `options` function returns and the beginning of the `run` function I would be very grateful.

(I know it is very noobish but for me `url = 'haha' % (quote('test'),quote('testbis'))` is a wrong expression. And if `options` is a function that requires `self` argument, why do we write `options['voice']` which is for me a snytax used for dictionaries :/)

Cheers and sorry for the noobish questions, which may not have their place here...

**Comment by Archibalde** [comment]
Oh ok thanks I get it.

**Comment by XunZhan** [comment]
Hi Dador, I tried the installation step as mentioned but failed. It said there was something wrong with the code so I checked the files. I wonder if this is only working on Anki 2.0.x? Mine is Anki 2.1.x so it fails.

**Comment by Geniusssmit** [comment]
> @Imrith Yes:
> 
> ```python
>         try:
>             from urllib.parse import quote
>         except ImportError:
>             from urllib2 import quote
> ```
> 
> [Dador#1](https://github.com/Dador/awesometts-anki-addon/pull/1) added it without the backward compatibility because the other services don't have it

where do I need to insert this code? And should i grab "try:" (first string)

**Comment by Geniusssmit** [comment]
> Does this still work for 2.1? I copied the appropriate lines of code into my files, but get an error ”No module named 'urllib2’" when trying to add audio.

I have the same issue. How did you solved it?

**Comment by lucwastiaux** [comment]
@Dador could you check the "allow edits from maintainers" checkbox ? thank you. https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by Dador** [comment]
@lucwastiaux seems like it was already checked.
image

**Comment by lucwastiaux** [comment]
@Dador thanks the effort that went into this. I'm reviewing and will merge soon. To be fair to the service, I believe we should encourage users to get their own API keys, so I will probably add an API key field. But just curious, how did you extract the API key used by the android app ? Did you have to get the paid android app ?
Thank you.

**Comment by lucwastiaux** [comment]
@Dador thanks a lot for putting this together. Forvo will probably become one of our high quality services which covers a wide number of voices given its crow-sourced nature. I checked out your branch and made the following changes:

- Allow selecting sex (male/female)
- Allow selecting country of origin of the speaker, or ANY if you just want the best rated sound
- Allow selecting API URL. For my key which is the entry level plan, the "apifree" URL is the correct one

I also added some automated tests in various languages. Merged.

**Comment by lucwastiaux** [comment]
I just created an "ankiaddon" file for 1.16 which contains Forvo support, does anyone want to give it a try before I push this version onto AnkiWeb ? https://github.com/AwesomeTTS/awesometts-anki-addon/releases

**Comment by Dador** [comment]
@lucwastiaux, Yes, I think it's good idea to encourage users to get their own API keys.
Extracted API key was from free official Android application. It was a long time ago and seems official Android application became paid.

Thank you for adding Forvo support and continuing development of this addon.

**Comment by lucwastiaux** [comment]
@Dador I just release v1.16 on AnkiWeb with Forvo support, thank you. You can still use your old key, if you select the "corporateapi" URL.

**Review by omdenis — COMMENTED** [review]
Awesome solution!

+LIne 70
        from urllib.request import quote

to support Python3
